web/src/settings_exports.ts

```diff
@@ -248,3 +409,67 @@ export function set_up(): void {
         });
     });
 }
+
+function maybe_store_export_consent_data_and_return(export_consent: ExportConsent): boolean {
+    // Handles a race where the client has requested the server for export_consents
+    // to populate 'Export permissions' table but hasn't received the response yet,
+    // but received a few updated events which should be applied on top of the received
+    // response to avoid outdated table.
+    // We store the export_consent data received via events to apply them on top of
+    // the received response.
+    if (export_consents === undefined) {
+        queued_export_consents.push(export_consent);
+        return true;
+    }
+    return false;
+}
+
+export function add_export_consent_data_and_redraw(export_consent: ExportConsent): void {
+    if (!meta.loaded) {
+        return;
+    }
+
+    if (maybe_store_export_consent_data_and_return(export_consent)) {
+        return;
+    }
+
+    export_consents.push(export_consent);
+    redraw_export_consents_list();
+}
+
+export function remove_export_consent_data_and_redraw(user_id: number): void {
+    if (!meta.loaded) {
+        return;
+    }
+
+    if (export_consents === undefined) {
+        queued_export_consents.push(user_id);
+        return;
+    }
+
+    const index = export_consents.findIndex((export_consent) => export_consent.user_id === user_id);
+    if (index !== -1) {
+        export_consents.splice(index, 1);
+        redraw_export_consents_list();
+    }
+}
+
+export function update_export_consent_data_and_redraw(export_consent: ExportConsent): void {
+    if (!meta.loaded) {
+        return;
+    }
+
+    if (maybe_store_export_consent_data_and_return(export_consent)) {
+        return;
+    }
+
+    const index = export_consents.findIndex((item) => item.user_id === export_consent.user_id);
+    if (index !== -1) {
+        // consent toggled
+        export_consents[index]!.consented = export_consent.consented;
+    } else {
+        // user reactivated
+        export_consents.push(export_consent);
+    }
```

Comment:
This logic also gets a lot cleaner if using a `Map`.

Reply:
Done